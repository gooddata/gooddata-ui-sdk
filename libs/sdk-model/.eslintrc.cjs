// (C) 2020 GoodData Corporation
module.exports = {
    parser: "@typescript-eslint/parser",
    plugins: ["prettier", "sonarjs", "eslint-plugin-tsdoc", "regexp"],
    extends: [
        "@gooddata",
        "plugin:import-esm/recommended",
        "plugin:sonarjs/recommended",
        "plugin:regexp/recommended",
        "../../.eslintrc.js",
    ],
    parserOptions: { tsconfigRootDir: __dirname },
};
