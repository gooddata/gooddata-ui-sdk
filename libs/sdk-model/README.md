# GoodData.UI domain models

[![npm version](https://img.shields.io/npm/v/@gooddata/sdk-model)](https://www.npmjs.com/@gooddata/sdk-model)&nbsp;
[![npm monthly downloads](https://img.shields.io/npm/dm/@gooddata/sdk-model)](https://npmcharts.com/compare/@gooddata/sdk-model?minimal=true)&nbsp;
![typescript](https://img.shields.io/badge/typescript-first-blue?logo=typescript)

This package is a part of the [GoodData.UI SDK](https://sdk.gooddata.com/gooddata-ui/docs/about_gooddataui.html).
To learn more, check [the source monorepo](https://github.com/gooddata/gooddata-ui-sdk).

This package provides domain models for GoodData.UI.

These domain models are backend-agnostic. This makes them reusable across different [Analytical Backend](https://www.npmjs.com/package/@gooddata/sdk-backend-spi) implementations.
The package includes TypeScript type definitions, factory functions, functions to get or set certain properties
of the objects in an immutable way, and more. These are used in both the `@gooddata/sdk-backend-*` and `@gooddata/sdk-ui* packages`.

## License

(C) 2017-2022 GoodData Corporation

This project is under MIT License. See [LICENSE](https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-model/LICENSE).
