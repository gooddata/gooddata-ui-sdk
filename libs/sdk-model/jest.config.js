// (C) 2019 GoodData Corporation
import base from "../../common/config/jest/jest.config.base.mjs";

export default {
  ...base,
};