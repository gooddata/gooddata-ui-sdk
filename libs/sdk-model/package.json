{
    "name": "@gooddata/sdk-model",
    "version": "8.12.3-beta.0",
    "author": "GoodData",
    "description": "GoodData Model definitions used by UI components and Backend SPI and its implementations",
    "repository": {
        "type": "git",
        "url": "https://github.com/gooddata/gooddata-ui-sdk.git",
        "directory": "libs/sdk-model"
    },
    "license": "MIT",
    "main": "dist/index.js",
    "module": "esm/index.js",
    "es2015": "esm/index.js",
    "browser": "dist/index.js",
    "typings": "esm/index.d.ts",
    "sideEffects": false,
    "files": [
        "dist/**/*.js",
        "dist/**/*.json",
        "dist/**/*.d.ts",
        "dist/**/*.map",
        "esm/**/*.js",
        "esm/**/*.json",
        "esm/**/*.d.ts",
        "esm/**/*.map"
    ],
    "config": {
        "eslint": "-c .eslintrc.js --ext ts src/"
    },
    "scripts": {
        "clean": "rm -rf ci dist esm coverage *.log && jest --clearCache",
        "dev": "tsc -p tsconfig.dev.json --watch",
        "build": "concurrently \"npm run build-cjs\" \"npm run build-esm\" && npm run api-extractor",
        "build-cjs": "tsc -p tsconfig.build.json",
        "build-esm": "tsc -p tsconfig.build.esm.json",
        "api-extractor": "mkdir -p api && [ -z \"${CI}\" ] && (api-extractor run -l) || (api-extractor run)",
        "test": "jest --watch",
        "test-once": "jest --maxWorkers=${JEST_MAX_WORKERS:-'45%'}",
        "test-ci": "JEST_JUNIT_OUTPUT=./ci/results/test-results.xml jest --ci --config jest.ci.js",
        "eslint": "eslint $npm_package_config_eslint",
        "eslint-ci": "mkdir -p ./ci/results && eslint -f checkstyle -o ci/results/eslint-results.xml $npm_package_config_eslint",
        "prettier-check": "prettier --check '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "prettier-write": "prettier --write '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "dep-cruiser": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "dep-cruiser-ci": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "validate": "npm run dep-cruiser && npm run eslint && npm run prettier-check",
        "validate-ci": "npm run dep-cruiser-ci && npm run eslint-ci && npm run prettier-check"
    },
    "dependencies": {
        "json-stable-stringify": "^1.0.1",
        "lodash": "^4.17.19",
        "spark-md5": "^3.0.0",
        "stringify-object": "^3.3.0",
        "ts-invariant": "^0.7.3",
        "tslib": "^2.0.0"
    },
    "devDependencies": {
        "@gooddata/eslint-config": "^4.1.0",
        "@microsoft/api-documenter": "^7.17.0",
        "@microsoft/api-extractor": "^7.20.0",
        "@types/jest": "^27.0.1",
        "@types/json-stable-stringify": "^1.0.32",
        "@types/lodash": "^4.14.158",
        "@types/spark-md5": "^3.0.1",
        "@types/stringify-object": "^3.2.0",
        "@typescript-eslint/eslint-plugin": "^5.45.0",
        "@typescript-eslint/parser": "^5.45.0",
        "concurrently": "^6.0.2",
        "dependency-cruiser": "^12.12.2",
        "eslint": "^8.25.0",
        "eslint-plugin-header": "^3.0.0",
        "eslint-plugin-import": "^2.22.0",
        "eslint-plugin-jest": "^27.1.1",
        "eslint-plugin-no-only-tests": "^2.4.0",
        "eslint-plugin-prettier": "^4.0.0",
        "eslint-plugin-regexp": "^1.9.0",
        "eslint-plugin-sonarjs": "^0.16.0",
        "eslint-plugin-tsdoc": "^0.2.14",
        "jest": "^27.5.1",
        "jest-junit": "^3.0.0",
        "prettier": "~2.5.0",
        "ts-jest": "^27.0.5",
        "typescript": "4.0.2"
    }
}
