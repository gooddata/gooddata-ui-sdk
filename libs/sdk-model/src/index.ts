// (C) 2019-2021 GoodData Corporation
/**
 * @packageDocumentation
 */
export { DateAttributeGranularity, DateGranularity, AllTimeGranularity } from "./base/dateGranularities";

export {
    IAttribute,
    isAttribute,
    attributeLocalId,
    AttributePredicate,
    anyAttribute,
    idMatchAttribute,
    attributesFind,
    attributeUri,
    attributeIdentifier,
    attributeAlias,
    attributeDisplayFormRef,
} from "./execution/attribute";

export {
    newAttribute,
    modifyAttribute,
    AttributeBuilder,
    AttributeModifications,
    AttributeBuilderInput,
} from "./execution/attribute/factory";

export {
    ObjectType,
    Identifier,
    Uri,
    UriRef,
    IdentifierRef,
    LocalIdRef,
    ObjRef,
    ObjRefInScope,
    isUriRef,
    isIdentifierRef,
    objRefToString,
    isLocalIdRef,
    areObjRefsEqual,
    isObjRef,
} from "./objRef";

export {
    IDimension,
    isDimension,
    dimensionTotals,
    DimensionItem,
    newTwoDimensional,
    newDimension,
    MeasureGroupIdentifier,
    dimensionSetTotals,
    dimensionsFindItem,
    ItemInDimension,
} from "./execution/base/dimension";

export { idRef, uriRef, localIdRef } from "./objRef/factory";

export { TotalType, ITotal, isTotal, newTotal, totalIsNative } from "./execution/base/totals";

export {
    SortDirection,
    IAttributeSortItem,
    ISortItem,
    IMeasureSortItem,
    ILocatorItem,
    IAttributeLocatorItem,
    IMeasureLocatorItem,
    isMeasureLocator,
    isAttributeLocator,
    isMeasureSort,
    isAttributeSort,
    isAttributeAreaSort,
    newMeasureSort,
    newAttributeSort,
    newAttributeAreaSort,
    newAttributeLocator,
    SortEntityIds,
    sortEntityIds,
    sortDirection,
    sortMeasureLocators,
    attributeLocatorElement,
    attributeLocatorIdentifier,
    measureLocatorIdentifier,
} from "./execution/base/sort";

export {
    IAttributeElementsByRef,
    IAttributeElementsByValue,
    IAttributeElements,
    IPositiveAttributeFilter,
    INegativeAttributeFilter,
    IAbsoluteDateFilter,
    IRelativeDateFilter,
    IMeasureValueFilter,
    IRankingFilter,
    RankingFilterOperator,
    isRankingFilter,
    IFilter,
    INullableFilter,
    IMeasureFilter,
    IDateFilter,
    IAttributeFilter,
    isAbsoluteDateFilter,
    isRelativeDateFilter,
    isAllTimeDateFilter,
    attributeElementsIsEmpty,
    attributeElementsCount,
    isPositiveAttributeFilter,
    isNegativeAttributeFilter,
    isDateFilter,
    isMeasureValueFilter,
    ComparisonConditionOperator,
    IComparisonCondition,
    IRangeCondition,
    MeasureValueFilterCondition,
    RangeConditionOperator,
    isAttributeFilter,
    isAttributeElementsByRef,
    isAttributeElementsByValue,
    isComparisonCondition,
    isComparisonConditionOperator,
    isFilter,
    isRangeCondition,
    isRangeConditionOperator,
    filterIsEmpty,
    filterAttributeElements,
    filterMeasureRef,
    filterObjRef,
    IAbsoluteDateFilterValues,
    IRelativeDateFilterValues,
    absoluteDateFilterValues,
    relativeDateFilterValues,
    measureValueFilterCondition,
    measureValueFilterMeasure,
    measureValueFilterOperator,
} from "./execution/filter";

export {
    newAbsoluteDateFilter,
    newNegativeAttributeFilter,
    newPositiveAttributeFilter,
    newRelativeDateFilter,
    newAllTimeFilter,
    newMeasureValueFilter,
    newRankingFilter,
} from "./execution/filter/factory";

export { mergeFilters } from "./execution/filter/filterMerge";

export {
    IMeasureTitle,
    IMeasureDefinitionType,
    IMeasureDefinition,
    ArithmeticMeasureOperator,
    IArithmeticMeasureDefinition,
    IPoPMeasureDefinition,
    IMeasure,
    MeasureAggregation,
    IPreviousPeriodMeasureDefinition,
    IPreviousPeriodDateDataSet,
    isMeasure,
    isSimpleMeasure,
    isAdhocMeasure,
    isPoPMeasure,
    isPreviousPeriodMeasure,
    isArithmeticMeasure,
    isMeasureDefinition,
    isPoPMeasureDefinition,
    isPreviousPeriodMeasureDefinition,
    isArithmeticMeasureDefinition,
    measureLocalId,
    MeasurePredicate,
    anyMeasure,
    idMatchMeasure,
    measureDoesComputeRatio,
    measureItem,
    measureUri,
    measureIdentifier,
    measureMasterIdentifier,
    measureArithmeticOperands,
    measureAlias,
    measureTitle,
    measureArithmeticOperator,
    measureFormat,
    isMeasureFormatInPercent,
    measureAggregation,
    measureFilters,
    measurePopAttribute,
    measurePreviousPeriodDateDataSets,
    MeasureOrLocalId,
} from "./execution/measure";

export {
    IPreviousPeriodDateDataSetSimple,
    ArithmeticMeasureBuilder,
    MeasureBuilder,
    MeasureModifications,
    PoPMeasureBuilder,
    PreviousPeriodMeasureBuilder,
    MeasureBuilderBase,
    newMeasure,
    modifyMeasure,
    modifySimpleMeasure,
    modifyPopMeasure,
    modifyPreviousPeriodMeasure,
    newArithmeticMeasure,
    newPopMeasure,
    newPreviousPeriodMeasure,
    MeasureEnvelope,
    ArithmeticMeasureBuilderInput,
    PoPMeasureBuilderInput,
    PreviousPeriodMeasureBuilderInput,
} from "./execution/measure/factory";

export {
    IAttributeOrMeasure,
    IBucket,
    isBucket,
    idMatchBucket,
    anyBucket,
    MeasureInBucket,
    AttributeInBucket,
    newBucket,
    bucketIsEmpty,
    bucketAttributes,
    bucketAttribute,
    bucketAttributeIndex,
    bucketMeasure,
    bucketMeasureIndex,
    bucketMeasures,
    bucketTotals,
    bucketSetTotals,
    bucketItems,
    BucketPredicate,
    applyRatioRule,
    ComputeRatioRule,
    disableComputeRatio,
    BucketItemModifications,
    BucketItemReducer,
    bucketModifyItems,
    bucketItemReduce,
} from "./execution/buckets";

export {
    bucketsFind,
    bucketsMeasures,
    bucketsIsEmpty,
    bucketsAttributes,
    bucketsFindMeasure,
    bucketsById,
    bucketsFindAttribute,
    bucketsItems,
    bucketsTotals,
    bucketsModifyItem,
    bucketsReduceItem,
} from "./execution/buckets/bucketArray";

export { bucketItemLocalId } from "./execution/buckets/bucketItem";

export {
    IExecutionDefinition,
    IExecutionConfig,
    DimensionGenerator,
    defWithFilters,
    defFingerprint,
    defSetDimensions,
    defSetSorts,
    defTotals,
    defSetExecConfig,
    IPostProcessing,
    defSetPostProcessing,
} from "./execution/executionDefinition";

export {
    newDefForItems,
    newDefForBuckets,
    newDefForInsight,
    defWithDimensions,
    defWithSorting,
    defWithPostProcessing,
    defWithDateFormat,
    defWithExecConfig,
    defaultDimensionsGenerator,
    emptyDef,
} from "./execution/executionDefinition/factory";

export {
    GuidType,
    RgbType,
    IRgbColorValue,
    IColor,
    IColorPalette,
    IColorPaletteItem,
    IColorFromPalette,
    IRgbColor,
    isColorFromPalette,
    isRgbColor,
    colorPaletteItemToRgb,
    colorPaletteToColors,
} from "./colors";

export {
    IInsight,
    IInsightDefinition,
    IVisualizationClass,
    VisualizationProperties,
    IColorMappingItem,
    isInsight,
    insightRef,
    insightId,
    insightItems,
    insightMeasures,
    insightHasMeasures,
    insightAttributes,
    insightHasAttributes,
    insightHasDataDefined,
    insightProperties,
    insightBuckets,
    insightSorts,
    insightBucket,
    insightTitle,
    insightUri,
    insightIsLocked,
    insightCreated,
    insightUpdated,
    insightTotals,
    insightFilters,
    insightVisualizationUrl,
    insightSetFilters,
    insightSetBuckets,
    insightSetProperties,
    insightSetSorts,
    insightModifyItems,
    insightReduceItems,
    visClassUrl,
    visClassId,
    visClassUri,
} from "./insight";

export { newInsightDefinition, InsightDefinitionBuilder, InsightModifications } from "./insight/factory";

export { insightSanitize } from "./insight/sanitization";

export { factoryNotationFor } from "./execution/objectFactoryNotation";
