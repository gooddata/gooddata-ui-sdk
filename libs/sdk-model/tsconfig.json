{
    "compilerOptions": {
        "target": "es2017",
        "lib": ["es2017", "dom"],
        "module": "NodeNext",
        "moduleResolution": "nodenext",
        "esModuleInterop": true,
        "importHelpers": true,        
        "sourceMap": true,
        "outDir": "./esm",
        "strict": true,
        "noImplicitAny": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noImplicitReturns": true,
        "removeComments": false
    },

    "exclude": ["**/node_modules","esm"]
}
