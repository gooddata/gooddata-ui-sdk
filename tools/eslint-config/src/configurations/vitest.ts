// (C) 2025 GoodData Corporation

import type { IConfiguration } from "../types.js";

const configuration: IConfiguration<"@vitest"> = {
    packages: [
        {
            name: "@vitest/eslint-plugin",
            version: "1.3.5",
        },
    ],
    plugin: "@vitest",
    extends: ["plugin:@vitest/legacy-recommended"],
    rules: {
        "@vitest/expect-expect": "off",
        "@vitest/no-commented-out-tests": "warn",
        "@vitest/valid-title": "warn",
        "@vitest/no-disabled-tests": "warn",
        "@vitest/no-focused-tests": "warn",
        "@vitest/no-identical-title": "warn",
        "@vitest/valid-expect": "warn",
    },
};

export default configuration;
