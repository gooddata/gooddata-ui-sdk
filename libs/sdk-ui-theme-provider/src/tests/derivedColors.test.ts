// (C) 2021-2022 GoodData Corporation
import { ThemeColor, IThemePalette } from "@gooddata/sdk-model";
import { describe, it, expect } from "vitest";
import {
    getHigherContrastColor,
    getLowerContrastColor,
    getLeastContrastColor,
    mixWith0ComplementaryColor,
    mixWith8ComplementaryColor,
} from "../derivedColors.js";

describe("derivedColors", () => {
    describe("getHigherContrastColor", () => {
        const Scenarios: Array<[number, ThemeColor, boolean]> = [
            [0.2, "#7f7f7f", false],
            [0.2, "#7f7f7f", true],
        ];

        it.each(Scenarios)("should generate higher contrast color", (amount, color, isDarkTheme) => {
            expect(getHigherContrastColor(amount, color, isDarkTheme)).toMatchSnapshot();
        });
    });

    describe("getLowerContrastColor", () => {
        const Scenarios: Array<[number, ThemeColor, boolean]> = [
            [0.2, "#7f7f7f", false],
            [0.2, "#7f7f7f", true],
        ];

        it.each(Scenarios)("should generate lower contrast color", (amount, color, isDarkTheme) => {
            expect(getLowerContrastColor(amount, color, isDarkTheme)).toMatchSnapshot();
        });
    });

    describe("getLeastContrastColor", () => {
        const Scenarios: Array<[ThemeColor, boolean]> = [
            ["#f00", false],
            ["#f00", true],
        ];

        it.each(Scenarios)("should generate least contrast color", (color, isDarkTheme) => {
            expect(getLeastContrastColor(color, isDarkTheme)).toMatchSnapshot();
        });
    });

    describe("mixWith0ComplementaryColor", () => {
        const Scenarios: Array<[number, ThemeColor, IThemePalette]> = [
            [0.5, "#f00", { complementary: { c0: "#fff", c9: "#000" } }],
            [0.5, "#f00", { complementary: { c0: "#000", c9: "#fff" } }],
        ];

        it.each(Scenarios)(
            "should generate color mixed with first color from the complementary palette",
            (amount, color, palette) => {
                expect(mixWith0ComplementaryColor(amount, color, palette)).toMatchSnapshot();
            },
        );
    });

    describe("mixWith8ComplementaryColor", () => {
        const Scenarios: Array<[number, ThemeColor, IThemePalette]> = [
            [0.5, "#f00", { complementary: { c0: "#fff", c8: "#333", c9: "#000" } }],
            [0.5, "#f00", { complementary: { c0: "#000", c8: "#ccc", c9: "#fff" } }],
        ];

        it.each(Scenarios)(
            "should generate color mixed with 8th color from the complementary palette",
            (amount, color, palette) => {
                expect(mixWith8ComplementaryColor(amount, color, palette)).toMatchSnapshot();
            },
        );
    });
});
