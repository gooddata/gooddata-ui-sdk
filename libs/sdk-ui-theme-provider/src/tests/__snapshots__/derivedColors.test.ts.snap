// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`derivedColors > getHigherContrastColor > should generate higher contrast color 1`] = `"#4c4c4c"`;

exports[`derivedColors > getHigherContrastColor > should generate higher contrast color 2`] = `"#b2b2b2"`;

exports[`derivedColors > getLeastContrastColor > should generate least contrast color 1`] = `"#ffebeb"`;

exports[`derivedColors > getLeastContrastColor > should generate least contrast color 2`] = `"#140000"`;

exports[`derivedColors > getLowerContrastColor > should generate lower contrast color 1`] = `"#b2b2b2"`;

exports[`derivedColors > getLowerContrastColor > should generate lower contrast color 2`] = `"#4c4c4c"`;

exports[`derivedColors > mixWith0ComplementaryColor > should generate color mixed with first color from the complementary palette 1`] = `"#ff7f7f"`;

exports[`derivedColors > mixWith0ComplementaryColor > should generate color mixed with first color from the complementary palette 2`] = `"#7f0000"`;

exports[`derivedColors > mixWith8ComplementaryColor > should generate color mixed with 8th color from the complementary palette 1`] = `"#991919"`;

exports[`derivedColors > mixWith8ComplementaryColor > should generate color mixed with 8th color from the complementary palette 2`] = `"#e56666"`;
